common/crypto/platform/openssl/crypto.cpp

```diff
@@ -50,6 +55,30 @@ using namespace std;
 namespace error = mender::common::error;
 namespace io = mender::common::io;
 
+#ifdef MENDER_CRYPTO_OPENSSL_LEGACY
+using EnginePtr = unique_ptr<ENGINE, void (*)(ENGINE *)>;
+#else
+using ProviderPtr = unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)>;
+#endif // MENDER_CRYPTO_OPENSSL_LEGACY
+
+#ifdef MENDER_CRYPTO_OPENSSL_LEGACY
+class OpenSSLResourceHandle {
+public:
+	EnginePtr engine;
+};
+#else
+class OpenSSLResourceHandle {
+public:
+	ProviderPtr default_provider_;
+	ProviderPtr hsm_provider_;
+};
+#endif // MENDER_CRYPTO_OPENSSL_LEGACY
+auto resource_handle_free_func = [](OpenSSLResourceHandle *h) {
+	if (h) {
+		h->~OpenSSLResourceHandle();
```

Comment:
Need to use `delete h` here, to pair it with the `new` below, otherwise the memory of `OpenSSLResourceHandle` itself won't be freed.

But calling `delete` is the default, so actually, you don't need a custom deleter at all.

Reply:
Oh.. true.

But I do need a custom deleter, if I am to use this way of hiding the resource handle definition in the `cpp` file, for otherwise the unique_ptr requires knowledge of the `OpenSSLResourceHandle` struct.

So I guess question is. Is it enough to call delete on `h` ? Or do i also need to explictly call the destructor?

Reply:
Yes, `delete` is enough, it both calls the destructor, and frees the memory.

However, I suspect you will run into the same problem, and I have a hunch why: Since this is a simple object, the destructor is (implicitly) defined in the header, where the full type is not available. Try to declare `~OpenSSLResourceHandle();` in the header, and put an empty one in the cpp file, which I think should be enough to move its definition there, avoiding the need for a full type.

I think...

Reply:
Seems `delete` does both :raised_hands:  :)

Reply:
> However, I suspect you will run into the same problem, and I have a hunch why: Since this is a simple object, the destructor is (implicitly) defined in the header, where the full type is not available. Try to declare ~OpenSSLResourceHandle(); in the header, and put an empty one in the cpp file, which I think should be enough to move its definition there, avoiding the need for a full type.

> I think...

The compiler kept asking for `sizeof(OpenSSLResourceHandle)`, so I don't think we are as lucky :/